README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

![Bolder Flight Systems Logo](img/logo-words_75.png) &nbsp; &nbsp; ![Arduino Logo](img/arduino_logo_75.png)

# TruStabilityRsc
This library communicates with [Honeywell TruStability RSC](https://sps.honeywell.com/us/en/products/advanced-sensing-technologies/healthcare-sensing/board-mount-pressure-sensors/trustability-rsc-series) pressure transducers. These sensors come in a wide variety of pressure ranges, measure pressure with 24 bit analog to digital converters, and communicate over SPI. This library is compatible with Arduino and CMake build systems.
   * [License](LICENSE.md)
   * [Changelog](CHANGELOG.md)
   * [Contributing guide](CONTRIBUTING.md)

# Installation

## Arduino
Use the Arduino Library Manager to install this library or clone to your Arduino/libraries folder. This library is added as:

```C++
#include "tru-stability-rsc.h"
```

Example Arduino executables are located in: *examples/arduino/*. Teensy 3.x, 4.x, and LC devices are used for testing under Arduino and this library should be compatible with other Arduino devices.

## CMake
CMake is used to build this library, which is exported as a library target called *tru-stability-rsc*. This library is added as:

```C++
#include "tru-stability-rsc.h"
```

The library can be also be compiled stand-alone using the CMake idiom of creating a *build* directory and then, from within that directory issuing:

```
cmake .. -DMCU=MK66FX1M0
make
```

This will build the library and example executables called *spi_example*. The example executable source file is located at *examples/cmake*. Notice that the *cmake* command includes a define specifying the microcontroller the code is being compiled for. This is required to correctly configure the code, CPU frequency, and compile/linker options. The available MCUs are:
   * MK20DX128
   * MK20DX256
   * MK64FX512
   * MK66FX1M0
   * MKL26Z64
   * IMXRT1062_T40
   * IMXRT1062_T41
   * IMXRT1062_MMOD

These are known to work with the same packages used in Teensy products. Also switching packages is known to work well, as long as it's only a package change.

The example targets create executables for communicating with the sensor using SPI communication. Each target also has a *_hex*, for creating the hex file to upload to the microcontroller, and an *_upload* for using the [Teensy CLI Uploader](https://www.pjrc.com/teensy/loader_cli.html) to flash the Teensy. Please note that instructions for setting up your build environment can be found in our [build-tools repo](https://github.com/bolderflight/build-tools).

# Namespace
This library is within the namespace *bfs*.

# TruStabilityRsc

## Methods

**TruStabilityRsc()** Default constructor, requires calling the *Config* method to setup the SPI bus and chip select pin.

**TruStabilityRsc(SPIClass &ast;spi, const uint8_t cs_ee, const uint8_t cs_adc)** Creates a *TruStabilityRsc* object. A pointer to the SPI bus object is passed along with the chip select pin of the sensor EEPROM and the chip select pin of the sensor ADC.

```C++
/*
* TruStability RSC sensor on SPI with EEPROM CS on pin 20 and ADC CS on pin 19
*/
bfs::TruStabilityRsc pres(&SPI, 20, 19);
```

**void Config(SPIClass &ast;spi, const uint8_t cs_ee, const uint8_t cs_adc)** This is required when using the default constructor and sets up the SPI bus and chip select pins.

**bool Begin()** Initializes communication with the sensor. The communication bus is not initialized within this library and must be initialized seperately; this enhances compatibility with other sensors that may on the same bus.

```C++
/* Initialize sensor */
SPI.begin();
pres.Begin();
```

**void ConfigMode(const Mode rate)** Configures the sampling mode; whether the sensor is in normal or fast mode and the sampling rate.

| Enum | Description |
| --- | --- |
| NORMAL_MODE_20HZ | Normal mode, 20Hz |
| NORMAL_MODE_45HZ | Normal mode, 45Hz |
| NORMAL_MODE_90HZ | Normal mode, 90Hz |
| NORMAL_MODE_175HZ | Normal mode, 175Hz |
| NORMAL_MODE_330HZ | Normal mode, 330Hz |
| NORMAL_MODE_600HZ | Normal mode, 600Hz |
| NORMAL_MODE_1000HZ | Normal mode, 1000Hz |
| FAST_MODE_40HZ | Fast mode, 40Hz |
| FAST_MODE_90HZ | Fast mode, 90Hz |
| FAST_MODE_180HZ | Fast mode, 180Hz |
| FAST_MODE_350HZ | Fast mode, 350Hz |
| FAST_MODE_660HZ | Fast mode, 660Hz |
| FAST_MODE_1200HZ | Fast mode, 1200Hz |
| FAST_MODE_2000HZ | Fast mode, 2000Hz |

**Mode mode()** Returns the configured mode.

**void temp_srd(const uint8_t srd)** Configures the temperature sample rate divider. Die temperature is used to compensate the pressure measurement; however, die temperature typically changes at a much slower rate than pressure. This sample rate divider enables configuring how often a temperature measurement is taken. If the srd is set to 0, calling the *Read* method will alternate between taking pressure and die temperature measurements. Setting the srd to 1 means two pressure measurements will be taken before a temperature measurement.

```C++
pres.temp_srd(10);
```

**uint8_t temp_srd()** Returns the configured srd.

**const char * prod_name()** Returns the pressure transducer name.

**const char * serial_num()** Returns the pressure transducer serial number.

**void Read()** Reads the sensor and stores the latest pressure or temperature measurement in the object. Note that there is no feedback about whether the sensor data is new. The data ready interrupt pin can be used to indicate this or timing based on the configured sampling mode.

```C++
pres.Read();
```

**float pres_pa()** Returns the latest pressure measurement from the object, Pa.

```C++
pres.pres_pa();
```

**float die_temp_c()** Returns the latest die temperature measurement from the object, C.

```C++
pres.die_temp_c();
```


CHANGELOG.md
# Changelog

## v1.0.3
- Added missing Arduino example

## v1.0.2
- Updated core to v3.1.3

## v1.0.1
- Fixed bool return on Begin

## v1.0.0
- Initial baseline


examples/cmake/spi.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2022 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "tru-stability-rsc.h"


/*
* TruStability RSC sensor on SPI with EEPROM CS on pin 20 and ADC CS on pin 19
*/
bfs::TruStabilityRsc pres(&SPI, 20, 19);

int main() {
  /* Serial monitor for displaying data */
  Serial.begin(115200);
  while (!Serial) {}
  /* Initialize sensor */
  SPI.begin();
  pres.Begin();
  /* Set temperature SRD */
  pres.temp_srd(10);
  while (1) {
    /* Read and display data */
    pres.Read();
    Serial.print(pres.pres_pa());
    Serial.print("\t");
    Serial.println(pres.die_temp_c());
    delay(100);
  }
}



src/tru-stability-rsc.cpp
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2022 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "tru-stability-rsc.h"  // NOLINT

namespace bfs {

void TruStabilityRsc::Config(SPIClass *spi, const uint8_t cs_ee,
                             const uint8_t cs_adc) {
  spi_ = spi;
  cs_ee_ = cs_ee;
  cs_adc_ = cs_adc;
}

bool TruStabilityRsc::Begin() {
  /* Configure the CS pins */
  pinMode(cs_ee_, OUTPUT);
  pinMode(cs_adc_, OUTPUT);
  digitalWrite(cs_ee_, HIGH);
  digitalWrite(cs_adc_, HIGH);
  /* Get the product name */
  ReadEepromRegisters(PROD_NAME_REG_, PROD_NAME_LEN_, buf_);
  memcpy(prod_name_, buf_, PROD_NAME_LEN_);
  /* Get the serial number */
  ReadEepromRegisters(SERIAL_NUM_REG_, SERIAL_NUM_LEN_, buf_);
  memcpy(serial_num_, buf_, SERIAL_NUM_LEN_);
  /* Get the pressure range */
  ReadEepromRegisters(PRES_RANGE_REG_, PRES_RANGE_LEN_, buf_);
  memcpy(&pres_range_, buf_, PRES_RANGE_LEN_);
  /* Get the pressure minimum */
  ReadEepromRegisters(PRES_MIN_REG_, PRES_MIN_LEN_, buf_);
  memcpy(&pres_min_, buf_, PRES_MIN_LEN_);
  /* Get the pressure unit */
  ReadEepromRegisters(PRES_UNIT_REG_, PRES_UNIT_LEN_, buf_);
  memcpy(pres_unit_, buf_, PRES_UNIT_LEN_);
  for (int8_t i = 0; pres_unit_[i]; i++) {
    pres_unit_[i] = tolower(pres_unit_[i]);
  }
  if (strcmp(pres_unit_, "inh2o") == 0) {
    conv_to_pa_ = 248.84f;
  } else if (strcmp(pres_unit_, "psi") == 0) {
    conv_to_pa_ = 0.45359237f * 9.80665f / 0.0254f / 0.0254f;
  } else if (strcmp(pres_unit_, "mbar") == 0) {
    conv_to_pa_ = 100.0f;
  } else if (strcmp(pres_unit_, "bar") == 0) {
    conv_to_pa_ = 100000.0f;
  } else if (strcmp(pres_unit_, "pa") == 0) {
    conv_to_pa_ = 1.0f;
  } else if (strcmp(pres_unit_, "kpa") == 0) {
    conv_to_pa_ = 1000.0f;
  } else {
    return false;
  }
  /* Get the pressure unit reference */
  ReadEepromRegisters(PRES_UNIT_REF_REG_, PRES_UNIT_REF_LEN_, buf_);
  memcpy(pres_unit_ref_, buf_, PRES_UNIT_REF_LEN_);
  /* Get the ADC config settings */
  ReadEepromRegisters(ADC_CONFIG_0_REG_, ADC_CONFIG_LEN_, buf_);
  memcpy(&adc_config_[0], buf_, ADC_CONFIG_LEN_);
  ReadEepromRegisters(ADC_CONFIG_1_REG_, ADC_CONFIG_LEN_, buf_);
  memcpy(&adc_config_[1], buf_, ADC_CONFIG_LEN_);
  ReadEepromRegisters(ADC_CONFIG_2_REG_, ADC_CONFIG_LEN_, buf_);
  memcpy(&adc_config_[2], buf_, ADC_CONFIG_LEN_);
  ReadEepromRegisters(ADC_CONFIG_3_REG_, ADC_CONFIG_LEN_, buf_);
  memcpy(&adc_config_[3], buf_, ADC_CONFIG_LEN_);
  /* Get the polynomial coefficients */
  ReadEepromRegisters(OFFSET_COEFF_0_REG_, OFFSET_COEFF_LEN_, buf_);
  memcpy(&offset_matrix_[0], buf_, OFFSET_COEFF_LEN_);
  ReadEepromRegisters(OFFSET_COEFF_1_REG_, OFFSET_COEFF_LEN_, buf_);
  memcpy(&offset_matrix_[1], buf_, OFFSET_COEFF_LEN_);
  ReadEepromRegisters(OFFSET_COEFF_2_REG_, OFFSET_COEFF_LEN_, buf_);
  memcpy(&offset_matrix_[2], buf_, OFFSET_COEFF_LEN_);
  ReadEepromRegisters(OFFSET_COEFF_3_REG_, OFFSET_COEFF_LEN_, buf_);
  memcpy(&offset_matrix_[3], buf_, OFFSET_COEFF_LEN_);
  /* Get the span coefficients */
  ReadEepromRegisters(SPAN_COEFF_0_REG_, SPAN_COEFF_LEN_, buf_);
  memcpy(&span_matrix_[0], buf_, SPAN_COEFF_LEN_);
  ReadEepromRegisters(SPAN_COEFF_1_REG_, SPAN_COEFF_LEN_, buf_);
  memcpy(&span_matrix_[1], buf_, SPAN_COEFF_LEN_);
  ReadEepromRegisters(SPAN_COEFF_2_REG_, SPAN_COEFF_LEN_, buf_);
  memcpy(&span_matrix_[2], buf_, SPAN_COEFF_LEN_);
  ReadEepromRegisters(SPAN_COEFF_3_REG_, SPAN_COEFF_LEN_, buf_);
  memcpy(&span_matrix_[3], buf_, SPAN_COEFF_LEN_);
  /* Get the shape coefficients */
  ReadEepromRegisters(SHAPE_COEFF_0_REG_, SHAPE_COEFF_LEN_, buf_);
  memcpy(&shape_matrix_[0], buf_, SHAPE_COEFF_LEN_);
  ReadEepromRegisters(SHAPE_COEFF_1_REG_, SHAPE_COEFF_LEN_, buf_);
  memcpy(&shape_matrix_[1], buf_, SHAPE_COEFF_LEN_);
  ReadEepromRegisters(SHAPE_COEFF_2_REG_, SHAPE_COEFF_LEN_, buf_);
  memcpy(&shape_matrix_[2], buf_, SHAPE_COEFF_LEN_);
  ReadEepromRegisters(SHAPE_COEFF_3_REG_, SHAPE_COEFF_LEN_, buf_);
  memcpy(&shape_matrix_[3], buf_, SHAPE_COEFF_LEN_);
  /* Send ADC corrections */
  WriteAdcRegisters(ADC_CONF_REG_, 4, adc_config_);
  /* Grab a temperature sample */
  TempDataConv();
  delay(65);
  ReadData(3, buf_);
  temp_c_ = ConvertTemp(3, buf_);
  PresDataConv();
  delay(65);
  ReadData(3, buf_);
  pres_ = ConvertPres(3, buf_);
  pres_pa_ = pres_ * conv_to_pa_;
  return true;
}

void TruStabilityRsc::ConfigMode(const Mode rate) {
  mode_ = static_cast<uint8_t>(rate);
  TempDataConv();
  delay(65);
  ReadData(3, buf_);
  temp_c_ = ConvertTemp(3, buf_);
  PresDataConv();
  delay(65);
  ReadData(3, buf_);
  pres_ = ConvertPres(3, buf_);
  pres_pa_ = pres_ * conv_to_pa_;
}

float TruStabilityRsc::ConvertTemp(const uint8_t count, uint8_t * const data) {
  traw_ = (static_cast<uint16_t>(data[0]) << 8 | data[1]) >> 2;
  if (traw_ & 0x2000) {
    tcomp_ = traw_ << 2;
    tcomp_ = ~tcomp_;
    tcomp_ = tcomp_ >> 2;
    tcomp_++;
    return static_cast<float>(tcomp_) * -0.03125f;
  } else {
    return static_cast<float>(traw_) * 0.03125f;
  }
}

float TruStabilityRsc::ConvertPres(const uint8_t count, uint8_t * const data) {
  praw_ = static_cast<int32_t>(data[0]) << 24 |
          static_cast<int32_t>(data[1]) << 16  |
          static_cast<int32_t>(data[2]) << 8;
  praw_ /= 256;
  prawf_ = static_cast<float>(praw_);
  trawf_ = static_cast<float>(traw_);
  pint1_ = prawf_ - (offset_matrix_[3] * trawf_ * trawf_ * trawf_ +
                     offset_matrix_[2] * trawf_ * trawf_ +
                     offset_matrix_[1] * trawf_ +
                     offset_matrix_[0]);
  pint2_ = pint1_ / (span_matrix_[3] * trawf_ * trawf_ * trawf_ +
                     span_matrix_[2] * trawf_ * trawf_ +
                     span_matrix_[1] * trawf_ +
                     span_matrix_[0]);
  pcompfs_ = shape_matrix_[3] * pint2_ * pint2_ * pint2_ +
             shape_matrix_[2] * pint2_ * pint2_ +
             shape_matrix_[1] * pint2_ +
             shape_matrix_[0];
  return pcompfs_ * pres_range_ + pres_min_;
}

void TruStabilityRsc::Read() {
  ReadData(3, buf_);
  if (meas_ == TEMP) {
    temp_c_ = ConvertTemp(3, buf_);
    PresDataConv();
  } else {
    pres_ = ConvertPres(3, buf_);
    pres_pa_ = pres_ * conv_to_pa_;
    cnt_++;
  }
  if (cnt_ > srd_) {
    cnt_ = 0;
    TempDataConv();
  }
}

void TruStabilityRsc::ReadData(const uint8_t count, uint8_t * const data) {
  spi_->beginTransaction(SPISettings(SPI_CLOCK_, MSBFIRST, SPI_MODE1));
  #if defined(TEENSYDUINO)
  digitalWriteFast(cs_adc_, LOW);
  #else
  digitalWrite(cs_adc_, LOW);
  #endif
  spi_->transfer(DATA_CONV_START_);
  spi_->transfer(data, count);
  #if defined(TEENSYDUINO)
  digitalWriteFast(cs_adc_, HIGH);
  #else
  digitalWrite(cs_adc_, HIGH);
  #endif
  spi_->endTransaction();
}

void TruStabilityRsc::TempDataConv() {
  reg_msb_ = mode_ | TEMP_READING_;
  WriteAdcRegisters(ADC_USR_CONF_REG_, 1, &reg_msb_);
  meas_ = TEMP;
}

void TruStabilityRsc::PresDataConv() {
  reg_msb_ = mode_ | PRES_READING_;
  WriteAdcRegisters(ADC_USR_CONF_REG_, 1, &reg_msb_);
  meas_ = PRES;
}

void TruStabilityRsc::ResetAdc() {
  spi_->beginTransaction(SPISettings(SPI_CLOCK_, MSBFIRST, SPI_MODE1));
  #if defined(TEENSYDUINO)
  digitalWriteFast(cs_adc_, LOW);
  #else
  digitalWrite(cs_adc_, LOW);
  #endif
  spi_->transfer(ADC_RESET_);
  #if defined(TEENSYDUINO)
  digitalWriteFast(cs_adc_, HIGH);
  #else
  digitalWrite(cs_adc_, HIGH);
  #endif
  spi_->endTransaction();
}

void TruStabilityRsc::WriteAdcRegisters(const uint16_t addr,
                                        const uint8_t count,
                                        uint8_t * const data) {
  reg_lsb_ = WREG_ | ((addr << 2) & ADC_REG_MASK_) |
             ((count - 1) & ADC_COUNT_MASK_);
  spi_->beginTransaction(SPISettings(SPI_CLOCK_, MSBFIRST, SPI_MODE1));
  #if defined(TEENSYDUINO)
  digitalWriteFast(cs_adc_, LOW);
  #else
  digitalWrite(cs_adc_, LOW);
  #endif
  spi_->transfer(reg_lsb_);
  spi_->transfer(data, count);
  #if defined(TEENSYDUINO)
  digitalWriteFast(cs_adc_, HIGH);
  #else
  digitalWrite(cs_adc_, HIGH);
  #endif
  spi_->endTransaction();
}

void TruStabilityRsc::ReadEepromRegisters(const uint16_t addr,
                                          const uint8_t count,
                                          uint8_t * const data) {
  reg_msb_ = EAD_EEPROM_ | ((addr & EEPROM_MASK_) >> 5);
  reg_lsb_ = addr & 0xFF;
  spi_->beginTransaction(SPISettings(SPI_CLOCK_, MSBFIRST, SPI_MODE0));
  #if defined(TEENSYDUINO)
  digitalWriteFast(cs_ee_, LOW);
  #else
  digitalWrite(cs_ee_, LOW);
  #endif
  spi_->transfer(reg_msb_);
  spi_->transfer(reg_lsb_);
  spi_->transfer(data, count);
  #if defined(TEENSYDUINO)
  digitalWriteFast(cs_ee_, HIGH);
  #else
  digitalWrite(cs_ee_, HIGH);
  #endif
  spi_->endTransaction();
}

}  // namespace bfs


src/tru-stability-rsc.h
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2022 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#ifndef TRU_STABILITY_RSC_SRC_TRU_STABILITY_RSC_H_  // NOLINT
#define TRU_STABILITY_RSC_SRC_TRU_STABILITY_RSC_H_

#if defined(ARDUINO)
#include <Arduino.h>
#include "SPI.h"
#else
#include <cstddef>
#include <cstdint>
#include "core/core.h"
#endif

namespace bfs {

class TruStabilityRsc {
 public:
  enum Mode : uint8_t {
    NORMAL_MODE_20HZ =    0b00000100,
    NORMAL_MODE_45HZ =    0b00100100,
    NORMAL_MODE_90HZ =    0b01000100,
    NORMAL_MODE_175HZ =   0b01100100,
    NORMAL_MODE_330HZ =   0b10000100,
    NORMAL_MODE_600HZ =   0b10100100,
    NORMAL_MODE_1000HZ =  0b11000100,
    FAST_MODE_40HZ =      0b00010100,
    FAST_MODE_90HZ =      0b00110100,
    FAST_MODE_180HZ =     0b01010100,
    FAST_MODE_350HZ =     0b01110100,
    FAST_MODE_660HZ =     0b10010100,
    FAST_MODE_1200HZ =    0b10110100,
    FAST_MODE_2000HZ =    0b11010100
  };
  TruStabilityRsc() {}
  TruStabilityRsc(SPIClass *spi, const uint8_t cs_ee, const uint8_t cs_adc) :
                  spi_(spi), cs_ee_(cs_ee), cs_adc_(cs_adc) {}
  void Config(SPIClass *spi, const uint8_t cs_ee, const uint8_t cs_adc);
  bool Begin();
  void Read();
  void ConfigMode(const Mode rate);
  inline void temp_srd(const uint8_t srd) {srd_ = srd;}
  inline uint8_t temp_srd() const {return srd_;}
  inline Mode mode() const {return static_cast<Mode>(mode_);}
  inline const char * prod_name() const {return prod_name_;}
  inline const char * serial_num() const {return serial_num_;}
  inline float pres_pa() const {return pres_pa_;}
  inline float die_temp_c() const {return temp_c_;}

 private:
  SPIClass *spi_;
  uint8_t cs_ee_, cs_adc_;
  static constexpr int32_t SPI_CLOCK_ = 5000000;
  static constexpr uint8_t EAD_EEPROM_ = 0x03;
  static constexpr uint8_t WREG_ = 0x40;
  static constexpr uint16_t EEPROM_MASK_ = 0x100;
  static constexpr uint8_t ADC_REG_MASK_ = 0x0C;
  static constexpr uint8_t ADC_COUNT_MASK_ = 0x03;
  uint8_t reg_lsb_, reg_msb_;
  /* Registers */
  static constexpr uint16_t PROD_NAME_REG_ = 0;
  static constexpr uint8_t PROD_NAME_LEN_ = 16;
  static constexpr uint16_t SERIAL_NUM_REG_ = 16;
  static constexpr uint8_t SERIAL_NUM_LEN_ = 11;
  static constexpr uint16_t PRES_RANGE_REG_ = 27;
  static constexpr uint8_t PRES_RANGE_LEN_ = 4;
  static constexpr uint16_t PRES_MIN_REG_ = 31;
  static constexpr uint8_t PRES_MIN_LEN_ = 4;
  static constexpr uint16_t PRES_UNIT_REG_ = 35;
  static constexpr uint8_t PRES_UNIT_LEN_ = 5;
  static constexpr uint16_t PRES_UNIT_REF_REG_ = 40;
  static constexpr uint8_t PRES_UNIT_REF_LEN_ = 1;
  static constexpr uint16_t ADC_CONFIG_0_REG_ = 61;
  static constexpr uint16_t ADC_CONFIG_1_REG_ = 63;
  static constexpr uint16_t ADC_CONFIG_2_REG_ = 65;
  static constexpr uint16_t ADC_CONFIG_3_REG_ = 67;
  static constexpr uint8_t ADC_CONFIG_LEN_ = 1;
  static constexpr uint16_t OFFSET_COEFF_0_REG_ = 130;
  static constexpr uint16_t OFFSET_COEFF_1_REG_ = 134;
  static constexpr uint16_t OFFSET_COEFF_2_REG_ = 138;
  static constexpr uint16_t OFFSET_COEFF_3_REG_ = 142;
  static constexpr uint8_t OFFSET_COEFF_LEN_ = 4;
  static constexpr uint16_t SPAN_COEFF_0_REG_ = 210;
  static constexpr uint16_t SPAN_COEFF_1_REG_ = 214;
  static constexpr uint16_t SPAN_COEFF_2_REG_ = 218;
  static constexpr uint16_t SPAN_COEFF_3_REG_ = 222;
  static constexpr uint8_t SPAN_COEFF_LEN_ = 4;
  static constexpr uint16_t SHAPE_COEFF_0_REG_ = 290;
  static constexpr uint16_t SHAPE_COEFF_1_REG_ = 294;
  static constexpr uint16_t SHAPE_COEFF_2_REG_ = 298;
  static constexpr uint16_t SHAPE_COEFF_3_REG_ = 302;
  static constexpr uint8_t SHAPE_COEFF_LEN_ = 4;
  static constexpr uint8_t ADC_RESET_ = 0x06;
  static constexpr uint8_t ADC_CONF_REG_ = 0x00;
  static constexpr uint8_t ADC_USR_CONF_REG_ = 0x01;
  static constexpr uint8_t DATA_CONV_START_ = 0x10;
  static constexpr uint8_t TEMP_READING_ = 0x02;
  static constexpr uint8_t PRES_READING_ = 0x00;
  /* Config */
  uint8_t mode_ = static_cast<uint8_t>(NORMAL_MODE_20HZ);
  uint8_t srd_ = 0, cnt_ = 0;
  uint8_t buf_[16];
  char prod_name_[PROD_NAME_LEN_ + 1] = {'\0'};
  char serial_num_[SERIAL_NUM_LEN_ + 1] = {'\0'};
  char pres_unit_[PRES_UNIT_LEN_ + 1] = {'\0'};
  char pres_unit_ref_[PRES_UNIT_REF_LEN_ + 1] = {'\0'};
  uint8_t adc_config_[4];
  float pres_range_;
  float pres_min_;
  float offset_matrix_[4];
  float span_matrix_[4];
  float shape_matrix_[4];
  enum Meas {
    TEMP,
    PRES
  } meas_;
  /* Data */
  uint16_t traw_, tcomp_;
  int32_t praw_;
  float trawf_, prawf_;
  float temp_c_;
  float pint1_, pint2_, pcompfs_;
  float pres_, pres_pa_;
  float conv_to_pa_;
  /* Utility functions */
  float ConvertTemp(const uint8_t count, uint8_t * const data);
  float ConvertPres(const uint8_t count, uint8_t * const data);
  void TempDataConv();
  void PresDataConv();
  void ReadData(const uint8_t count, uint8_t * const data);
  void ResetAdc();
  void WriteAdcRegisters(const uint16_t addr, const uint8_t count,
                         uint8_t * const data);
  void ReadEepromRegisters(const uint16_t addr, const uint8_t count,
                           uint8_t * const data);
};

}  // namespace bfs

#endif  // TRU_STABILITY_RSC_SRC_TRU_STABILITY_RSC_H_ NOLINT
